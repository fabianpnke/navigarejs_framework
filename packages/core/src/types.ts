import Route from './Route'
import { GeneratedRoutes } from './routes'
import { Default, Wildcard } from './symbols'
import { AxiosResponse, Canceler, CancelToken } from 'axios'
import { IsEmptyObject } from 'type-fest'

// Router
export type RouterOptions<TComponent> = {
  initialPage: Page
  base: string
  resolveComponentModule?: ComponentModuleResolver<TComponent>
  fragments?: Record<
    string,
    {
      stacked?: boolean
      lazy?: boolean
    }
  >
  events?: {
    onBefore?: RouterEventListener<'before'>
    onStart?: RouterEventListener<'start'>
    onProgress?: RouterEventListener<'progress'>
    onFinish?: RouterEventListener<'finish'>
    onCancel?: RouterEventListener<'cancel'>
    onSuccess?: RouterEventListener<'success'>
    onError?: RouterEventListener<'error'>
    onInvalid?: RouterEventListener<'invalid'>
    onException?: RouterEventListener<'exception'>
  }
}

export type ComponentModuleResolver<TComponent> = (
  url: string,
  component: PageComponent,
) => Promise<TComponent>

export type RouterLocation = {
  href: string
  host: string
  hostname: string
  origin: string
  pathname: string
  port: string
  protocol: string
  search: string
  hash: string
  state?: string | undefined
}

// Pages
export type DeferredValue = {
  __deferred: true
}

export type DeferredProperty<TValue = any> = DeferredValue | TValue

export type PageDefaults = Record<string, any>

export type PageErrors = Record<string, string[]>

export type PageErrorBag = Record<string, PageErrors>

export interface PageProperties {
  [key: string]: unknown
}

export type PageRememberedState = Record<string, unknown> | undefined

export type PageFragmentProperties = Record<string, any>

export type PageComponent = {
  id: string
  path: string
}

export type PageFragment = {
  component: PageComponent
  properties: PageFragmentProperties
  /*rawRoute: RawRoute
  location: RouterLocation
  defaults: PageDefaults
  parameters: Record<string, RouteParameter>*/
  page?: Page
}

export type PageFragments = Record<string, PageFragment | PageFragment[] | null>

export interface Page {
  visit: Visit
  csrf: string | null
  fragments: PageFragments
  properties: PageProperties & {
    errors?: PageErrors & PageErrorBag
  }
  rawRoute: RawRoute
  location: RouterLocation
  defaults: PageDefaults
  parameters: Record<string, RouteParameter>
  version: string | null
  layout: string | null
  timestamp: number
  base?: Page

  obsolete: boolean
  scrollRegions: Array<{ top: number; left: number }>
  rememberedState: PageRememberedState
}

// Events
export type RouterEvents = {
  before: {
    details: {
      visit: Visit
    }
    result: boolean | void
  }

  start: {
    parameters: [Visit]
    details: {
      visit: Visit
    }
    result: void
  }

  progress: {
    details: {
      visit: Visit
      progress: VisitProgress | undefined
    }
    result: void
  }

  finish: {
    details: {
      visit: Visit
    }
    result: void
  }

  cancel: {
    details: {
      visit: Visit
    }
    result: void
  }

  navigate: {
    details: {
      visit: Visit
      page: Page
      pages: Page[]
      pageIndex: number
      replace: boolean
    }
    result: void
  }

  success: {
    details: {
      visit: Visit
      page: Page
    }
    result: void
  }

  error: {
    details: {
      visit: Visit
      errors: PageErrors
    }
    result: void
  }

  invalid: {
    details: {
      visit: Visit
      response: AxiosResponse
    }
    result: boolean | void
  }

  exception: {
    details: {
      visit: Visit
      error: Error
    }
    result: boolean | void
  }
}

export type RouterEventNames = keyof RouterEvents

export type RouterEventDetails<TEventName extends RouterEventNames> =
  RouterEvents[TEventName]['details']

export type RouterEventResult<TEventName extends RouterEventNames> =
  RouterEvents[TEventName]['result']

export type RouterEventListener<TEventName extends RouterEventNames> = (
  event: CustomEvent<RouterEventDetails<TEventName>>,
) => RouterEventResult<TEventName>

// Visits
export type FormDataConvertible =
  | Array<FormDataConvertible>
  | Blob
  | FormDataEntryValue
  | Date
  | boolean
  | number
  | null
  | Record<string, any>

export type VisitData = Record<string, FormDataConvertible> | FormData

export type VisitCancelToken = {
  token?: CancelToken
  cancel: Canceler
}

export type VisitPreserveStateOption =
  | boolean
  | string
  | ((page: Page) => boolean)

export type VisitProgress = {
  loaded: number
  total?: number
  progress?: number
  bytes: number
  rate?: number
  estimated?: number
  upload?: boolean
  download?: boolean
}

export type LocationVisit = {
  preserveScroll: boolean
}

export type VisitOptions = Partial<{
  method: RawRouteMethod
  data: VisitData
  replace: boolean
  preserveScroll: VisitPreserveStateOption
  preserveState: VisitPreserveStateOption
  properties: Array<string>
  headers: Record<string, string>
  errorBag: string | null
  forceFormData: boolean
  queryStringArrayFormat: QueryStringArrayFormat
  onBefore: RouterEventListener<'before'>
  onStart: RouterEventListener<'start'>
  onProgress: RouterEventListener<'progress'>
  onFinish: RouterEventListener<'finish'>
  onCancel: RouterEventListener<'cancel'>
  onSuccess: RouterEventListener<'success'>
  onError: RouterEventListener<'error'>
  onInvalid: RouterEventListener<'invalid'>
  onException: RouterEventListener<'exception'>
}>

export type Visit = {
  id: VisitId
  method: RouteMethod
  data: VisitData
  replace: boolean
  preserveScroll: VisitPreserveStateOption
  preserveState: VisitPreserveStateOption
  properties: Array<string>
  headers: Record<string, string>
  errorBag: string | null
  forceFormData: boolean
  queryStringArrayFormat: QueryStringArrayFormat
  location: RouterLocation
  completed: boolean
  cancelled: boolean
  interrupted: boolean
  cancelToken?: VisitCancelToken
  cancel?: () => void
  interrupt?: () => void
  onBefore?: RouterEventListener<'before'>
  onStart?: RouterEventListener<'start'>
  onProgress?: RouterEventListener<'progress'>
  onFinish?: RouterEventListener<'finish'>
  onCancel?: RouterEventListener<'cancel'>
  onSuccess?: RouterEventListener<'success'>
  onError?: RouterEventListener<'error'>
  onInvalid?: RouterEventListener<'invalid'>
  onException?: RouterEventListener<'exception'>
}

export type VisitId = string

export type Component = unknown

// Routes
export type Routes = GeneratedRoutes & {}

export enum RouteMethod {
  GET = 'GET',
  POST = 'POST',
  PUT = 'PUT',
  PATCH = 'PATCH',
  DELETE = 'DELETE',
}

export type RouteName = keyof Routes

export type Routable<TRouteName extends RouteName = RouteName> =
  | URL
  | string
  | Route<TRouteName>

export interface BindableRouteParameter {
  id: number | string
}

export type RouteParameter =
  // | BindableRouteParameter
  string | number | boolean | Record<string, any> | undefined | null

/*export type RouteParameters<TParamValue = RouteParameter> = {
  [key: string]: TParamValue
}*/

export type RouteParameters<TRouteName extends RouteName> =
  Routes[TRouteName] extends {
    bindings: any
  }
    ? IsEmptyObject<Routes[TRouteName]['bindings']> extends true
      ? Record<never, never>
      : Record<keyof Routes[TRouteName]['bindings'], RouteParameter>
    : Record<never, never>

export type RouteBindings = Record<string, string | null>

export type RouteWheres = Record<string, string | null>

export type RouteDefaults = Record<string, RouteParameter>

export type RawRouteParameters<TRouteName extends RouteName = RouteName> =
  Routes[TRouteName] extends {
    bindings: any
  }
    ? IsEmptyObject<Routes[TRouteName]['bindings']> extends true
      ? Record<never, never>
      : Record<
          keyof Routes[TRouteName]['bindings'],
          RouteParameter | typeof Default | typeof Wildcard
        >
    : Record<never, never>

export type RawRouteMethod =
  | 'GET'
  | 'HEAD'
  | 'POST'
  | 'PATCH'
  | 'PUT'
  | 'DELETE'

export type RawRoute<TName extends RouteName = RouteName> = {
  name: TName
  uri: string
  methods: RawRouteMethod[]
  method?: RawRouteMethod
  domain?: string
  bindings?: RouteBindings
  parameters?: RouteParameters<TName>
  wheres?: RouteWheres
  components?: PageComponent[]
}

export type RawRoutes = Record<string, RawRoute>

export enum QueryStringArrayFormat {
  Indices = 'indices',
  Brackets = 'brackets',
}

// Rendered
export interface RenderedApp {
  id: string
  modules: Set<string>
  headTags: string
  htmlAttributes: string
  bodyAttributes: string
  bodyTags: string
  appHTML: string
}
