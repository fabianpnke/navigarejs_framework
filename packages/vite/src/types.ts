import { RawRoutes } from '@navigare/core'
import { Plugin as VitePlugin } from 'vite'

export type Plugin = VitePlugin & {}

export enum Adapter {
  Laravel,
}

export type Configuration = {
  ssr: {
    enabled: boolean
    protocol: string
    port: number
    host: string
    timeout: number
    input: string
    manifest: string
  }
  client: {
    manifest: string
  }
  components: {
    path: string
    default_extension: string
  }
  testing: {
    ensure_pages_exist: boolean
  }
}

export interface CommandsConfiguration {
  artisan: Record<string, string[]> | string[]
  shell: Record<string, string[]> | string[]
}

export interface Options {
  /**
   * Source of routes and configuration.
   */
  routes?: Adapter | RawRoutes | string

  /**
   * Interval that determines how often the routes will be refreshed.
   */
  interval?: number | false

  /**
   * Path to PHP executable.
   */
  php?: string

  /**
   * A configuration object or a path to a configuration file.
   * Setting to false disables reading the configuration file path from the `CONFIG_PATH_NAVIGARE` environment variable.
   */
  configuration?: Adapter | Configuration | string

  /**
   * Whether to allow overrides from the base configuration. If false, base
   * options will be ignored, so stuff like `--host 0.0.0.0` won't work.
   *
   * @default true
   */
  allowOverrides?: boolean

  /**
   * List of file changes to listen to.
   */
  watch?: WatchInput[] | WatchOptions
}

export interface WatchOptions {
  reloadOnBladeUpdates?: boolean
  reloadOnControllerUpdates?: boolean
  reloadOnConfigUpdates?: boolean
  input?: WatchInput[]
}

export interface WatchInputHandlerParameters {
  file: string
}

export interface WatchInput {
  condition: (file: string) => boolean
  handle: (parameters: WatchInputHandlerParameters) => void
}

export interface PhpFinderOptions {
  /**
   * Actual path to PHP. This will be used instead of
   */
  path?: string

  /**
   * Custom environment variables.
   */
  env?: any

  /**
   * Either `production` or `development`. Used for loading the environment.
   */
  mode?: string
}
