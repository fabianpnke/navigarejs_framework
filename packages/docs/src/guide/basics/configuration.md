# Configuration

You can publish the configuration for Laravel at any time via the following command:

```bash
php artisan vendor:publish --provider="Navigare\ServiceProvider"
```

This will copy the default configuration to `config/navigare.php` and it will allow you to fine tune some default settings. For example, if you don't want your components to reside in `resources/scripts/pages` then you can define a different path.
