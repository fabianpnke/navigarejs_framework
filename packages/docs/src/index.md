---
layout: home

hero:
  name: Navigare
  tagline: Quickly build modern single-page apps with Vue and Laravel.
  image:
    src: https://raw.githubusercontent.com/navigarejs/framework/main/assets/logo.svg
    alt: Navigare
  actions:
    - theme: brand
      text: Get Started
      link: /guide/getting-started/introduction.html
    - theme: alt
      text: View on GitHub
      link: https://github.com/navigarejs/framework
---
