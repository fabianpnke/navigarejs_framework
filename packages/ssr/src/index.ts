export { default as serve, default as serveSSR } from './serve'

export * from './types'
