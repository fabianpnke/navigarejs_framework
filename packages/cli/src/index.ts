import { getVersion } from './internals.js'
import { serve } from '@navigare/ssr'
import * as commander from 'commander'
import path from 'path'

const program = new commander.Command().version(getVersion())

program
  .command('serve')
  .description('Run server and render pages.')
  .argument('<string>', 'path to input file', (file) => {
    return path.join(process.cwd(), file)
  })
  .option('-m, --manifest <char>', 'path to manifest file', (file) => {
    return path.join(process.cwd(), file)
  })
  .option('-p, --port <number>', 'port number', '13714')
  .action(async function (input) {
    const { port, manifest } = this.opts()

    const { printUrls } = await serve({
      input,
      manifest,
      port: Number(port),
      logger: {
        info: (message) => {
          console.info(message)
        },
        warn: (message) => {
          console.warn(message)
        },
        warnOnce: (message) => {
          console.warn(message)
        },
        error: (message) => {
          console.error(message)
        },
      },
    })

    printUrls()
  })

export default program
