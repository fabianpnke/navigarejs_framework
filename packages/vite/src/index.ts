import { getConfiguration } from './configuration'
import { getRoutes, writeTypes } from './routes'
import { Options, Adapter, Configuration, Plugin } from './types'
import _generate from '@babel/generator'
import { parse } from '@babel/parser'
import _traverse from '@babel/traverse'
import type { NodePath } from '@babel/traverse'
import {
  isImportSpecifier,
  isIdentifier,
  isCallExpression,
  identifier,
  isStringLiteral,
  isMemberExpression,
  CallExpression,
  callExpression,
  Expression,
  ArgumentPlaceholder,
  JSXNamespacedName,
  SpreadElement,
  Program,
  importDeclaration,
  importSpecifier,
  stringLiteral,
} from '@babel/types'
import { RawRoute, RawRoutes, throwError } from '@navigare/core'
import { Server as SSRServer, serveSSR } from '@navigare/ssr'
import crypto from 'crypto'
import makeDebugger from 'debug'
import getPort from 'get-port'
import globby from 'globby'
import cloneDeep from 'lodash.clonedeep'
import defaultsDeep from 'lodash.defaultsdeep'
import get from 'lodash.get'
import isArray from 'lodash.isarray'
import isObject from 'lodash.isobject'
import isString from 'lodash.isstring'
import { createRequire } from 'module'
import path from 'path'
import { ConfigEnv, createLogger, createServer, loadEnv, Logger } from 'vite'

// Annoying workaround to circumvent ESM/CJS confusion
// See: https://github.com/babel/babel/issues/13855#issuecomment-945123514
const traverse = (
  'default' in _traverse ? get(_traverse, 'default') : _traverse
) as typeof _traverse
const generate = (
  'default' in _generate ? get(_generate, 'default') : _generate
) as typeof _generate

const require = createRequire(import.meta.url)

const debug = makeDebugger('navigare:laravel:index')

export default function createNavigarePlugin(options: Options = {}): Plugin {
  defaultsDeep(options, {
    routes: Adapter.Laravel,
    configuration: Adapter.Laravel,
    interval: 15000,
  })
  let currentRoutes: RawRoutes = isObject(options.routes) ? options.routes : {}
  let currentConfiguration: Configuration | null = isObject(
    options.configuration,
  )
    ? options.configuration
    : null
  let currentComponents: string[] = []
  let ssr = false
  let ssrServer: SSRServer | undefined = undefined
  let logger: Logger | undefined = undefined
  let environment: ConfigEnv = {
    mode: 'production',
    command: 'build',
  }
  const currentUsedRoutes: Record<string, string[]> = {}
  const files: Record<string, string> = {}
  const buildId = Math.random().toString()
  const getChunkName = (name: string) => {
    const hash = crypto
      .createHash('md5')
      .update(`${buildId}-${name}`)
      .digest('hex')

    return `${hash}.mjs`
  }

  return {
    name: 'navigare',

    enforce: 'post',

    buildEnd() {
      // Nothing yet
    },

    closeBundle() {
      // Nothing yet
    },

    async config(configuration, _environment) {
      environment = _environment

      // Create logger
      logger = createLogger(configuration.logLevel, {
        allowClearScreen: configuration.clearScreen,
        customLogger: configuration.customLogger,
        prefix: 'navigare',
      })

      // Loads .env
      const env = loadEnv(environment.mode, process.cwd(), '')

      // Set SSR mode
      ssr = !!configuration.build?.ssr
      debug('set SSR mode to %s', ssr)

      // Read routes regularly in case no routes were provided
      const updateRoutes = async () => {
        try {
          currentRoutes = await getRoutes(options, env, ssr)
          debug('read routes: %O', currentRoutes)

          // Write types
          if (currentRoutes) {
            // await writeTypes(currentConfiguration?.types.path, currentRoutes)
            await writeTypes(
              path.join(
                require.resolve('@navigare/core'),
                '..',
                '..',
                'types',
                'routes.d.ts',
              ),
              currentRoutes,
            )
          }
        } catch (error) {
          if (error instanceof Error) {
            logger?.error(error.message)
          }
        }

        // Update routes in a while again
        const interval = Number(options.interval)
        if (environment.command === 'serve' && interval > 0) {
          setTimeout(updateRoutes, interval)
        }
      }
      await updateRoutes()

      // Read configuration regularly in case no configuration was provided
      const updateConfiguration = async () => {
        try {
          currentConfiguration = await getConfiguration(options, env)
          debug('read configuration: %O', currentConfiguration)
        } catch (error) {
          if (error instanceof Error) {
            logger?.error(error.message)
          }
        }

        // Update configuration in a while again
        const interval = Number(options.interval)
        if (environment.command === 'serve' && interval > 0) {
          setTimeout(updateConfiguration, interval)
        }
      }
      await updateConfiguration()

      // Read components regularly
      const updateComponents = async () => {
        try {
          currentComponents = await globby(
            `${currentConfiguration?.components.path}/**/*`,
          )
          debug('read components: %O', currentComponents)
        } catch (error) {
          if (error instanceof Error) {
            logger?.error(error.message)
          }
        }

        // Update components in a while again
        const interval = Number(options.interval)
        if (environment.command === 'serve' && interval > 0) {
          setTimeout(updateComponents, interval)
        }
      }
      await updateComponents()

      // Enable manifest and extend input with all the pages that were used
      const input = ssr
        ? currentConfiguration?.ssr.input
        : configuration.build?.rollupOptions?.input
      configuration.build = {
        ...configuration.build,
        ssr: ssr,
        rollupOptions: {
          ...configuration.build?.rollupOptions,
          input: Object.values({
            // Cast existing input into object
            ...(isArray(input)
              ? input.reduce((cumulatedInput, name) => {
                  return {
                    ...cumulatedInput,
                    [name]: name,
                  }
                }, {})
              : isObject(input)
              ? input
              : isString(input)
              ? {
                  [input]: input,
                }
              : {}),
          }),
          output: {
            ...configuration.build?.rollupOptions?.output,
            chunkFileNames: (chunkInfo) => {
              if (
                chunkInfo.type !== 'chunk' ||
                !chunkInfo.isEntry ||
                !chunkInfo.facadeModuleId
              ) {
                return '[name]-[hash].mjs'
              }

              const id = path.relative(process.cwd(), chunkInfo.facadeModuleId)
              return getChunkName(id)
            },
          },
        },
        manifest: true,
        // ssrManifest: true,
      }

      // Update SSR config
      if (ssr) {
        configuration.ssr = configuration.ssr || {}
        configuration.ssr.external = [
          ...(configuration.ssr?.external || []),

          ...(environment.command === 'serve'
            ? [
                // Core
                ...[
                  'lodash.clonedeep',
                  'lodash.debounce',
                  'lodash.get',
                  'lodash.isarray',
                  'lodash.isequal',
                  'lodash.isboolean',
                  'lodash.isfunction',
                  'lodash.isnumber',
                  'lodash.isobject',
                  'lodash.isstring',
                  'lodash.issymbol',
                  'lodash.merge',
                  'lodash.uniq',
                ],
                // Vue3
                ...['vue', 'lodash.isequal'],
              ]
            : []),
        ]
      }
    },

    async configureServer(server) {
      // In case we are already in SSR mode, we don't need to start the server
      if (ssr) {
        return
      }

      // Create SSR server
      const vite = await createServer({
        build: {
          ssr: true,
        },
        server: {
          middlewareMode: true,
          hmr: {
            port: await getPort(),
          },
        },
        appType: 'custom',
      })

      // Start SSR server
      debug('serving SSR at %s', currentConfiguration?.ssr.port)
      ssrServer = await serveSSR({
        logger,
        port: currentConfiguration?.ssr.port,
        vite,
      })

      server.httpServer?.once('listening', () => {
        setTimeout(() => {
          ssrServer?.printUrls()
        }, 25)
      })
    },

    /*moduleParsed(module) {
      if (!module.isEntry) {
        return
      }

       module.dynamicImporters.push([
        "resources/scripts/pages/Home.vue",
       ]) 
    },*/

    buildStart() {
      // Include all possible components in final build
      if (environment.command === 'build') {
        for (const component of currentComponents) {
          files[component] = this.emitFile({
            type: 'chunk',
            id: component,
            preserveSignature: 'exports-only',
          })
        }
      }
    },

    async transform(code: string, id: string) {
      let updatedCode = code
      const usedRoutes: string[] = []

      // Only replace routes when the source code contains "route" and is not a module from node_modules
      if (code.includes('route') && !id.includes('node_modules')) {
        let manipulated = false

        // Parse code
        const ast = parse(updatedCode, {
          sourceType: 'module',
          plugins: ['jsx', 'typescript'],
        })

        const replaceArgumentWithRoute = (routeCall: CallExpression) => {
          if (!isStringLiteral(routeCall.arguments[0])) {
            /*console.log(routeCall)
          throwError(
            'the first parameter of "route" function must be a string literal',
            routeCall.arguments,
          )*/
            return
          }

          // Extract route name
          const routeName = routeCall.arguments[0].value
          const rawRoute: RawRoute | undefined = cloneDeep(
            currentRoutes?.[routeName],
          )
          if (!rawRoute) {
            throwError(`"${routeName}" is not a valid route name`)
          }

          // Update path of used components based on the build manifest
          if (environment.command === 'build' && isObject(rawRoute)) {
            rawRoute.components = rawRoute.components?.map((component) => {
              return {
                ...component,
                path: getChunkName(component.path),
              }
            })
          }

          // Replace route name with the raw route
          routeCall.arguments[0] = identifier(JSON.stringify(rawRoute))

          // Remember that we used the route and also indicate that we manipulated the AST
          usedRoutes.push(routeName)
          manipulated = true
        }
        const replaceWithRouteCall = (
          program: NodePath<Program>,
          path: NodePath,
          args: (
            | Expression
            | ArgumentPlaceholder
            | JSXNamespacedName
            | SpreadElement
          )[],
        ) => {
          const routeIdentifier = program.scope.generateUidIdentifier('route')
          const [importDeclarationPath] = program.unshiftContainer(
            'body',
            importDeclaration(
              [importSpecifier(routeIdentifier, identifier('route'))],
              stringLiteral('@navigare/core'),
            ),
          )
          program.scope.registerDeclaration(importDeclarationPath)

          const binding = program.scope.getBinding(routeIdentifier.name)
          const [routeCallExpressionPath] = path.replaceWith(
            callExpression(routeIdentifier, args),
          )
          binding?.reference(routeCallExpressionPath.get('callee'))

          // Remember that we changed something, so we can output the new formatted AST later
          manipulated = true
        }

        traverse(ast, {
          Program(path) {
            const program = path

            // Turn special cases into route(...) calls:
            // - _ctx.route(...)
            // - $setup.route(...)
            // - __unref(route)(...)
            const objectIdentifierNames = ['_ctx', '$setup']
            const calleeIdentifierNames = ['_unref']
            const propertyIdentifierNames = ['route']

            traverse(ast, {
              CallExpression(path) {
                const { callee } = path.node

                // Member expressions like _ctx.route(...) or $setup.route(...)
                if (isMemberExpression(callee)) {
                  if (!isIdentifier(callee.object)) {
                    return
                  }

                  if (!isIdentifier(callee.property)) {
                    return
                  }

                  if (!objectIdentifierNames.includes(callee.object.name)) {
                    return
                  }

                  if (!propertyIdentifierNames.includes(callee.property.name)) {
                    return
                  }

                  replaceWithRouteCall(program, path, path.node.arguments)
                  return
                }

                // Call expressions like __unref(route)(...)
                if (isCallExpression(callee)) {
                  if (!isIdentifier(callee.callee)) {
                    return
                  }

                  if (!calleeIdentifierNames.includes(callee.callee.name)) {
                    return
                  }

                  if (!isIdentifier(callee.arguments[0])) {
                    return
                  }

                  if (
                    !propertyIdentifierNames.includes(callee.arguments[0].name)
                  ) {
                    return
                  }

                  replaceWithRouteCall(program, path, path.node.arguments)
                  return
                }
              },
            })

            // Catch function calls like route(...)
            traverse(ast, {
              ImportDeclaration(path) {
                const { specifiers, source } = path.node

                if (source.value !== '@navigare/core') {
                  return
                }

                // Check if the route function was imported
                const routeIdentifiers = specifiers
                  .map((specifier) => {
                    if (!isImportSpecifier(specifier)) {
                      return null
                    }

                    if (!isIdentifier(specifier.imported)) {
                      return null
                    }

                    if (specifier.imported.name !== 'route') {
                      return null
                    }

                    return specifier.local.name
                  })
                  .filter((identifier): identifier is string => !!identifier)

                for (const routeIdentifier of routeIdentifiers) {
                  const binding = program.scope?.getBinding(routeIdentifier)
                  for (const reference of binding?.referencePaths ?? []) {
                    if (!isCallExpression(reference.parent)) {
                      return
                    }

                    replaceArgumentWithRoute(reference.parent)
                  }
                }
              },
            })
          },
        })

        if (manipulated) {
          updatedCode = generate(ast, {}).code
        }
      }

      // Remember all used routes per module
      currentUsedRoutes[id] = usedRoutes

      return updatedCode
    },
  }
}
