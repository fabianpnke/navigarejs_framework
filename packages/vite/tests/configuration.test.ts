import { getConfiguration } from '../src/configuration'
import { Adapter } from '../src/types'
import { artisan, fixture } from './utilities'
import { it, expect } from 'vitest'

it('reads configuration from the given option object', async () => {
  const configuration = await getConfiguration(
    {
      configuration: {
        ssr: {
          enabled: false,
          protocol: 'http',
          host: '127.0.0.1',
          port: 13715,
          timeout: 1,
          input: 'resources/scripts/ssr.ts',
          manifest: '/tmp/bootstrap/ssr/manifest.json',
        },

        client: {
          manifest: '/tmp/public/build/manifest.json',
        },

        components: {
          path: 'resources/scripts/pages',
          default_extension: 'vue',
        },

        testing: {
          ensure_pages_exist: true,
        },
      },
    },
    process.env,
  )

  expect(configuration).toMatchSnapshot()
})

it('reads configuration from file', async () => {
  const configuration = await getConfiguration(
    {
      configuration: fixture('configuration.json'),
    },
    process.env,
  )

  expect(configuration).toMatchSnapshot()
})

it('reads configuration from Laravel', async () => {
  process.env.TEST_ARTISAN_SCRIPT = artisan()

  const configuration = await getConfiguration(
    {
      configuration: Adapter.Laravel,
    },
    process.env,
  )

  expect(configuration).toMatchSnapshot()
})
