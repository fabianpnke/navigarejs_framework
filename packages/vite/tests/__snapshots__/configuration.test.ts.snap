// Vitest Snapshot v1

exports[`reads configuration from Laravel 1`] = `
{
  "ssr": {
    "enabled": false,
    "url": "http://127.0.0.1:13715/",
  },
  "testing": {
    "ensure_pages_exist": true,
    "page_extensions": [],
    "page_paths": [],
  },
  "types": {
    "path": "./resources/js/navigare.d.ts",
  },
}
`;

exports[`reads configuration from file 1`] = `
{
  "ssr": {
    "enabled": false,
    "url": "http://127.0.0.1:13715/",
  },
  "testing": {
    "ensure_pages_exist": true,
    "page_extensions": [],
    "page_paths": [],
  },
  "types": {
    "path": "./resources/js/navigare.d.ts",
  },
}
`;

exports[`reads configuration from the given option object 1`] = `
{
  "client": {
    "manifest": "/tmp/public/build/manifest.json",
  },
  "components": {
    "default_extension": "vue",
    "path": "resources/scripts/pages",
  },
  "ssr": {
    "enabled": false,
    "host": "127.0.0.1",
    "input": "resources/scripts/ssr.ts",
    "manifest": "/tmp/bootstrap/ssr/manifest.json",
    "port": 13715,
    "protocol": "http",
    "timeout": 1,
  },
  "testing": {
    "ensure_pages_exist": true,
  },
}
`;
